waspc/src/Wasp/Analyzer/TypeChecker/TypeError.hs

```diff
@@ -1,38 +1,52 @@
+{-# LANGUAGE LambdaCase #-}
+
 module Wasp.Analyzer.TypeChecker.TypeError
   ( TypeError (..),
-    TypeCoerceReason (..),
+    TypeCoercionErrorReason (..),
+    TypeCoercionError (..),
+    getErrorMessageAndCtx,
   )
 where
 
--- TODO:
--- TypeErrors are not detailed enough for the final version, missing is:
--- 1. Reporting multiple type errors:
---     This can be incrementally improved, e.g.
---     1. improve to 1 error per statement
---     2. improve to multiple errors per list/per dictionray
--- 2. Position information:
---     The start position in the source and end position should be recorded. (3) may
---     affect this, since it may be hard to say exactly where a unification or
---     weaken error happened.
--- 3. Informative error messages:
---     Make user-readable messages from all errors, and special care into explaining
---     unification and weaken errors (users shouldn't have to know what unification
---     and weakening is).
-
+import Control.Arrow (first)
+import Data.List (intercalate)
+import Wasp.Analyzer.Parser.Ctx (Ctx)
 import Wasp.Analyzer.Type
 import Wasp.Analyzer.TypeChecker.AST
+import Wasp.Util (concatPrefixAndText, concatShortPrefixAndText, indent)
 
+{- ORMOLU_DISABLE -}
 data TypeError
-  = UnificationError TypeCoerceReason Type Type
-  | WeakenError TypeCoerceReason TypedExpr Type
-  | NoDeclarationType String
-  | UndefinedIdentifier String
-  | QuoterUnknownTag String
-  | DictDuplicateField String
+  = UnificationError     Ctx TypeCoercionError
```

Comment:
Here, for some reason I didn't use WithCtx but instead embedded Ctx directly. I fixed this in the later PR (https://github.com/wasp-lang/wasp/pull/408) and replaced this with WithCtx.